### DISCUSSION TITLE
u::Lux AddOn (temp/humidity/motion/CO2 sensor) support for Tasmota

### DISCUSSION THREAD

**Discussion by KingAndi** [arendst/Tasmota#18862] [Ideas]
Dear Tasmota community!

My Name is Andreas König and I am the CEO of the u::Lux GmbH. We are the manufacturer of the [u::Lux Switch](https://www.u-lux.com/luxswitch) (IP-based TFT Display) and optional sensors for temp, humidity, motion and CO2.

I don't want to make advertising for our products in this post, but we came up with an idea to connect our sensors (called [u::Lux AddOns](https://www.u-lux.com/addons)) to tasmota. The sensors have a dimension of 55 to 55mm and fit in multiple cover frame systems like Gira, Jung, and many more. Our sensors are ready to use products including a housing, instead of DiY products. The sensors are powered with 3.3V and have a 100kBit I2C interface. Attached to the sensor is a 4 pin cable  (VCC, GND, SCL, SDA) with a molex connector.

At this point we would ask for some feedback from the community to know if there is a possibility to implement the I2C protocol to tasmota and if there is a general interest for the usage of our sensors in tasmota. The next step would be to find somebody who is interested to implement the necessary functions & features...

Best regards,
Andreas König

**Comment by barbudor** [comment]
Hi Andreas,

Generally there are no difficulties to integrate I2C sensors in Tasmota. We have already quite a lot of similar sensors.
In most of cases, this starts with someone purchasing such device and then having the skills to code the driver.
Some questions that comes to my mind:
- Is there only 1 I2C address and are the various sensors combinations hidden behind an aggregation MCU ?
- Or each sensor is a separated chip so each add-on model would have a different set of individual chips/addresses ?

If you want to do it by yourself, you can ask questions here. We would advise you on some existing drivers that you can take as a reference and general guidelines to lead to a successful PR.

If you prefer that someone do the coding on your behalf, then this you can ask to Theo @arendst if he have time for that or who could do it (on my side, I'm unfortunately too much busy with real-life work currently).

Is the u::Lux Switch based on ESP32 ? 
Do you have already considered running Tasmota on it ? May be it's already the case ?
With LVGL that could also be a nice fit.

Best regards

**Comment by KingAndi** [comment]
Hi Barbudor!

Thank you for your quick response. 

At the moment we have 5 different sensors:

- AddOn RH-Temp (one I2C address for temperature & humidity). All other AddOns use the same chip (HTU21D) for temperature and humidity measurement.
- AddOn RH-Temp-Motion (one I2C for temp & humidity). And a second I2C address for a digital i/o chip to poll the motion sensor and to control the motion detector RGB-LED (absence, presence or any other usage)
- AddOn RH-Temp-CO2 (one I2C for temp & humidity). And a second I2C address for the CO2 value (in ppm).
- AddOn Multisensor (completely new) which combines all functions, but it has a different CO2 sensor (SCD40). So this sensor has 3 different I2C addresses (one for temperature & humidity, one for CO2 and one for digital I/O to poll the motion sensor and to control the motion detector LED and 3 additional RGB-LED's for CO2 state.
- AddOn IN2 (only two digital inputs). We don't think it makes sense to implement this too because most devices have already inputs...

We would prefer that somebody who knows Tasmota and its behaviour to implement the functions. Of course we give support with free AddOns and other products…

The u::Lux Switch itself is based on a LPC1768 (Cortex M3) and so it's not based on ESP32. But of course we are thinking about new technics like MQTT and maybe Tastmota too...

best regards,
Andreas

**Reply by s-hadinger** [reply]
If the controlling device is an ESP32x (down to ESP32C3), you can very easily create I2C driver in the Berry language. Some examples are here: https://tasmota.github.io/docs/Berry-Cookbook/#creating-an-i2c-driver

**Reply by sfromis** [reply]
When it comes to those addon modules, Tasmota is already supporting many I2C sensors, and if your addon devices are simply connecting the sensor chips directly to the external I2C bus, the existing support for HTU21 and SCD40 should work. If the PIR sensor has a direct external connection, Tasmota can also support that, but if it is "wrapped" in some protocol, things may get more complex. Of course, using such addons with Tasmota (without your switch device) would require an ESP chip "nearby"....

**Reply by barbudor** [reply]
if the IR sensor is connected to a PCF8574 or MCP230xx I/O extander, then this is already supported too.

So probably your Add-Ons are already covered.

**Reply by KingAndi** [reply]
Hallo!

I will write a single hardware manual with all the technical data to all sensors. I think this would be the best way...
At the moment with have internal manulas to each AddOn, but i cannot post this manuals here...

Best regards,
Andreas

**Comment by KingAndi** [comment]
Hallo! 

I am sorry for the delay, but i was on holiday... Thank you very much for your replies and your help. We have written a integration manual, which describes the different AddOns and which sensors each AddOns uses. There is some special to-do like initialization. At the moment the manual is only in German language, but in about 3 weeks it will be available in English language too. 

If anybody is interesting in integrating our AddOns in tasmota, please contact me, of cause we will supply with free samples!

Link to the manual: [https://www.u-lux.com/files/uLux_AddOns_Integration_Manual.pdf](https://www.u-lux.com/files/uLux_AddOns_Integration_Manual.pdf). If there are any questions or suggestions, please feel free to contact us too!

Short overview for the sensors and chips we use (details in the manual):

- HTU21D (Temperature & Humidity)
- PCA9536DR & PCA9539A for I/O’s (Motion detector and LED's)
- CDM7162 and SCD40 for CO2

Best regards,
Andreas

**Comment by KingAndi** [comment]
Hallo!

Had anybody time to look into the manual? Or do you need the manual in english language?
Looking forward for your replies!

Best regards,
Andreas

**Reply by HofeBY2** [reply]
As written above, developer need the englisch Version.

I read your Manual als german speeking, and made you an offer by EMail before your Hollydays.
But Im not a Tasmota developner !

I can show you how to set up compiling environment under Windows.
HTU21D and SCD40 are supportet by tasmota.
PCA9536DR and PCA9539A not currently.

I think it would be an good Idea set up an developmet environmet in Salzburg.

So long.

**Comment by KingAndi** [comment]
Hallo HofeBY2

Thanks for your reply and your offer, we will post the english manual next week and than talk again...

best regards,
Andreas

**Comment by KingAndi** [comment]
Hallo again!

Now we have translated the manual to English language. Please feel free to have a look into the manual. 
- German manual on GitHub [uLux_AddOns_Integration_Manual.pdf](https://github.com/arendst/Tasmota/files/12068910/uLux_AddOns_Integration_Manual.pdf)
Also available on our webserver [https://www.u-lux.com/files/uLux_AddOns_Integration_Manual.pdf](https://www.u-lux.com/files/uLux_AddOns_Integration_Manual.pdf)
- **English manual** on GitHub **[uLux_AddOns_Integration_Manual_en.pdf](https://github.com/arendst/Tasmota/files/12068906/uLux_AddOns_Integration_Manual_en.pdf)**
Also available on our webserver [https://www.u-lux.com/files/uLux_AddOns_Integration_Manual_en.pdf](https://www.u-lux.com/files/uLux_AddOns_Integration_Manual_en.pdf) or direct from

If anybody is interested in implementing the features to tasmota please contact Andreas König/u::Lux GmbH ([www.u-lux.com](https://www.u-lux.com/en/about-us/team/))

Best regards,
Andreas
